src/test/common_url_tests.cpp

```diff
@@ -54,6 +54,9 @@ BOOST_AUTO_TEST_CASE(decode_malformed_test) {
     BOOST_CHECK_EQUAL(UrlDecode(" %Z "), " %Z ");
     BOOST_CHECK_EQUAL(UrlDecode(" % X"), " % X");
 
+    BOOST_CHECK_EQUAL(UrlDecode("%%ffg"), "%\xffg");
```

Comment:
I had to scratch my head on this one...

It iterates from left to right, the first thing it finds is `%`. It then tries if the next two characters are hex, which they are not, so it puts the literal `%` in the result. Then it moves to the next `%` which is followed by valid hex `ff`. Since `0xff` is [invalid unicode](https://stackoverflow.com/questions/5657467/is-byte-0xff-valid-in-a-utf-8-encoded-string), we have to represent it with the [escape sequence](https://en.cppreference.com/w/cpp/language/escape) "\xff". And then there's the regular character `g`.